File: python/bot/cogs/info/whois.py
"""?whois — user profile with moderation history."""

from __future__ import annotations

import discord
from discord.ext import commands

from bot.cogs.deps import CogRepos
from bot.utils.resolvers import resolve_member
from bot.utils.time import format_iso_date


class WhoisCog(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot
        self.repos = CogRepos(bot)

    @commands.command(name="whois")
    async def whois(self, ctx: commands.Context[commands.Bot], *, user_arg: str | None = None) -> None:
        target = resolve_member(ctx.guild, ctx.message, user_arg)  # type: ignore[arg-type]
        if target is None and isinstance(ctx.author, discord.Member):
            target = ctx.author
        if target is None:
            await ctx.reply("Could not resolve that user.")
            return

        guild_id = str(ctx.guild.id)  # type: ignore[union-attr]
        warnings = await self.repos.warnings.list_active(guild_id, str(target.id))
        notes = await self.repos.notes.list_for_user(guild_id, str(target.id))
        cases = await self.repos.cases.list_for_user(guild_id, str(target.id), limit=5)

        embed = discord.Embed(title=str(target), color=target.colour.value or 0x5865F2)
        embed.set_thumbnail(url=target.display_avatar.url)
        embed.add_field(name="ID", value=str(target.id), inline=True)
        joined = f"<t:{int(target.joined_at.timestamp())}:R>" if target.joined_at else "Unknown"
        embed.add_field(name="Joined", value=joined, inline=True)
        embed.add_field(name="Created", value=f"<t:{int(target.created_at.timestamp())}:R>", inline=True)
        roles = [role.mention for role in target.roles if role != ctx.guild.default_role]  # type: ignore[union-attr]
        embed.add_field(name="Roles", value=" ".join(roles) or "None", inline=False)
        embed.add_field(name="Warnings", value=str(len(warnings)), inline=True)
        embed.add_field(name="Notes", value=str(len(notes)), inline=True)

        if cases:
            case_lines = [
                f"**#{case.case_number}** {case.action} — {case.reason or 'N/A'} ({format_iso_date(case.created_at)})"
                for case in cases
            ]
            embed.add_field(name="Recent Cases", value="\n".join(case_lines), inline=False)

        await ctx.reply(embed=embed)


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(WhoisCog(bot))

File: node-bun/src/commands/info/whois.js
import { EmbedBuilder } from 'discord.js';
import { resolveMember, error } from '../../utils/helpers.js';
import { getWarnings, getNotes, getCasesForUser } from '../../database/db.js';
import { formatDate } from '../../utils/time.js';

export default {
  name: 'whois',
  description: 'Get user information and mod history',
  category: 'Info',
  usage: '?whois [user]',
  async execute(message, args) {
    const target = resolveMember(message, args.trim()) ?? message.member;
    const user = target.user;

    const warnings = await getWarnings(message.guild.id, target.id);
    const notes = await getNotes(message.guild.id, target.id);
    const cases = await getCasesForUser(message.guild.id, target.id, 5);

    const embed = new EmbedBuilder()
      .setTitle(user.tag)
      .setThumbnail(user.displayAvatarURL({ size: 256 }))
      .setColor(target.displayHexColor || 0x5865f2)
      .addFields(
        { name: 'ID', value: user.id, inline: true },
        { name: 'Joined', value: target.joinedAt ? `<t:${Math.floor(target.joinedTimestamp / 1000)}:R>` : 'Unknown', inline: true },
        { name: 'Created', value: `<t:${Math.floor(user.createdTimestamp / 1000)}:R>`, inline: true },
        { name: 'Roles', value: target.roles.cache.filter((r) => r.id !== message.guild.id).map((r) => r.toString()).join(' ') || 'None' },
        { name: 'Warnings', value: `${warnings.length}`, inline: true },
        { name: 'Notes', value: `${notes.length}`, inline: true },
      );

    if (cases.length) {
      embed.addFields({
        name: 'Recent Cases',
        value: cases.map((c) => `**#${c.case_number}** ${c.action} — ${c.reason || 'N/A'} (${formatDate(c.created_at)})`).join('\n'),
      });
    }

    return message.reply({ embeds: [embed] });
  },
};
